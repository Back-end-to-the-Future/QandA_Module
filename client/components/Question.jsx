import React from 'react';
import axios from 'axios';

import Answer from './Answer.jsx';
import SubmitAnswer from './SubmitAnswer.jsx';

import Button from 'react-bootstrap/Button';
import Container from 'react-bootstrap/Container';
import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col';

class Question extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      question: this.props.question,
      answers: this.props.question.answers,
      isMoreAnswers: this.props.question.answers.isMoreAnswers,
      helpfulness: this.props.question.question_helpfulness,
      helpfulClicked: false,
      reported: false,
      showSubmitAnswer: false
    }

    this.incrementHelpfulQuestion = this.incrementHelpfulQuestion.bind(this);
    this.reportQuestion = this.reportQuestion.bind(this);
  }

  incrementHelpfulQuestion() {
    axios.put('/question/helpful', { question_id: this.state.question.question_id });
    this.setState({ helpfulness: this.state.helpfulness + 1, helpfulClicked: true });
  }

  reportQuestion() {
    axios.put('/question/report', { question_id: this.state.question.question_id })
      .catch((err) => {
        // commented for tests
        // un-comment to check errors
        // console.log(err);
      })

    this.setState({ reported: true });
  }

  getMoreAnswers() {
    axios.get('/moreAnswers', {
      params: {
        question_id: this.state.question.question_id
      }
    })
      .then((response) => {
        let answers = response.data.answers;
        let isMoreAnswers = response.data.isMoreAnswers;
        this.setState({ answers, isMoreAnswers });
      })
  }

  collapseAnswers() {
    this.setState({ answers: this.state.answers.slice(0, 2), isMoreAnswers: true });
  }

  handleOpenSubmit() {
    this.setState({ showSubmitAnswer: true });
  }

  handleCloseSubmit() {
    this.setState({ showSubmitAnswer: false });
  }

  render() {

    let answers;
    if (Array.isArray(this.state.answers)) {
      answers = this.state.answers;
    } else {
      let answersObj = this.state.answers;
      answers = [];
      for (let id in answersObj) {
        if (id !== 'isMoreAnswers') {
          answers.push(answersObj[id]);
        }
      }
    }

    // sort array of answers by helpfulness
    answers.sort((a, b) => (a.helpfulness > b.helpfulness) ? -1 : 1);

    // moves seller answers to top of the list
    for (let i = answers.length - 1; i >= 0; i--) {
      if (answers[i].answerer_name === 'Seller') {
        let sellerAns = answers.splice(i, 1);
        answers.unshift(sellerAns[0]);
      }
    }

    let moreAnswers;
    if (this.state.isMoreAnswers) {
      moreAnswers = <Button variant='outline-dark' className='moreAnswers' onClick={this.getMoreAnswers.bind(this)}>See More Answers</Button>;
    } else if (answers.length <= 2) {
      moreAnswers = <span></span>;
    } else {
      moreAnswers = <Button variant='outline-dark' className='collapseAnswers' onClick={this.collapseAnswers.bind(this)}>Collapse Answers</Button>;
    }

    let helpfulButton = <Button variant="outline-dark" size="sm" className="qHelpfulBtn"
      onClick={this.incrementHelpfulQuestion} >Yes ({this.state.helpfulness})</Button>;
    if (this.state.helpfulClicked) {
      helpfulButton = <Button variant="secondary" size="sm" className="qHelpfulBtn btn-secondary">Yes ({this.state.helpfulness})</Button>;
    }

    let reportButton = <Button variant="outline-dark" size="sm" className="qReportBtn" onClick={this.reportQuestion} >Report</Button>;
    if (this.state.reported) {
      reportButton = <Button variant="secondary" size="sm" className="qReportBtn btn-secondary">Reported</Button>;
    }

    return (
      <div>
        <Container className='question'>
          <Row>
            <Col xs='auto'><h5>Q: {this.state.question.question_body}</h5></Col>
          </Row>
          <Row>
            <Col className='qButtonRow'>
              <span>Helpful? </span> {helpfulButton}
              {reportButton}
            </Col>
          </Row>
          <Row className='answerRow'>
            <Col>
              {answers.map((answer, index) => {
                return (
                  <Answer answer={answer} key={index} />
                );
              })}
            </Col>
          </Row>
          <Row className='moreAnswersRow'>
            {moreAnswers}
          </Row>
          <Row className='submitAnswerRow'>
            <SubmitAnswer
              product_name={this.props.product_name}
              question_id={this.state.question.question_id}
              question_body={this.state.question.question_body}
              handleOpenSubmit={this.handleOpenSubmit.bind(this)}
              handleCloseSubmit={this.handleCloseSubmit.bind(this)}
              show={this.state.showSubmitAnswer}
            />
          </Row>
        </Container>
      </div>
    );
  }
}

export default Question;