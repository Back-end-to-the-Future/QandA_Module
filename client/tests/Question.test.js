import React from 'react';
import Question from '../components/Question.jsx';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

// Test Data for Question
const testQuestion = {"question_id":34,"question_body":"Can I wash it?","question_date":"2017-01-04T00:00:00.000Z","asker_name":"luaulover","question_helpfulness":23,"reported":0,"answers":{"10":{"id":10,"body":"I've thrown it in the wash and it seems fine","date":"2017-01-04T00:00:00.000Z","answerer_name":"skilover","helpfulness":9,"photos":["https://images.unsplash.com/photo-1510551310160-589462daf284?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=1649&q=80","https://images.unsplash.com/photo-1469504512102-900f29606341?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=668&q=80"]},"11":{"id":11,"body":"It says not to","date":"2017-01-04T00:00:00.000Z","answerer_name":"skilover","helpfulness":5,"photos":[]},"isMoreAnswers":true}};

describe('Question Component', () => {
  it('should change state.helpfulClicked to true and increment helpfulness on click of "Helpful? Yes" button', () => {
    const wrapper = shallow(<Question question={testQuestion} key={0}/>);
    let preClickHelpfulness = wrapper.state('helpfulness');

    let helpfulButton = wrapper.find('.qHelpfulButton');

    helpfulButton.simulate('click');

    let helpfulState = wrapper.state('helpfulClicked');
    let postClickHelpfulness = wrapper.state('helpfulness');

    expect(helpfulState).toEqual(true);
    expect(postClickHelpfulness).toEqual(preClickHelpfulness + 1);
  })

  it('should match test snapshot', () => {
    const questionDisplay = renderer.create(<Question question={testQuestion} key={0}/>).toJSON();
    expect(questionDisplay).toMatchSnapshot();
  })
});