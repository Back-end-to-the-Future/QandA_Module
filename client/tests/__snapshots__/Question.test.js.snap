// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Question Component should match test snapshot 1`] = `
<div
  className="questionDiv"
>
  <div
    className="question container"
  >
    <div
      className="row"
    >
      <div
        className="col-auto"
      >
        <h5>
          Q: 
          Can I wash it?
        </h5>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="qButtonRow col"
      >
        <span>
          Helpful? 
        </span>
         
        <button
          className="qHelpfulBtn btn btn-outline-dark btn-sm"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Yes (
          23
          )
        </button>
        <button
          className="qReportBtn btn btn-outline-dark btn-sm"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Report
        </button>
      </div>
    </div>
    <div
      className="answerRow row"
    >
      <div
        className="answerCol col"
      >
        <div
          className="answerContainer"
        >
          <span />
          <div
            className="answer container"
          >
            <div
              className="row"
            >
              <div
                className="answerBody col"
              >
                A: 
                I've thrown it in the wash and it seems fine
              </div>
            </div>
            <div
              className="row"
            >
              <div
                className="col"
              >
                by 
                skilover
                , 
                January 04, 2017
              </div>
            </div>
            <div
              className="thumbnailDiv row"
            >
              <div
                className="thumbnailContainer"
              >
                <img
                  className="thumbnail"
                />
              </div>
              <div
                className="thumbnailContainer"
              >
                <img
                  className="thumbnail"
                />
              </div>
            </div>
            <div
              className="row"
            >
              <div
                className="aButtonCol col"
              >
                <span>
                  Helpful? 
                </span>
                <button
                  className="aHelpfulBtn btn btn-outline-dark btn-sm"
                  disabled={false}
                  onClick={[Function]}
                  type="button"
                >
                  Yes (
                  9
                  )
                </button>
                <button
                  className="aReportBtn btn btn-outline-dark btn-sm"
                  disabled={false}
                  onClick={[Function]}
                  type="button"
                >
                  Report
                </button>
              </div>
            </div>
          </div>
          <div
            className="answer container"
          >
            <div
              className="row"
            >
              <div
                className="answerBody col"
              >
                A: 
                It says not to
              </div>
            </div>
            <div
              className="row"
            >
              <div
                className="col"
              >
                by 
                skilover
                , 
                January 04, 2017
              </div>
            </div>
            <div
              className="thumbnailDiv row"
            />
            <div
              className="row"
            >
              <div
                className="aButtonCol col"
              >
                <span>
                  Helpful? 
                </span>
                <button
                  className="aHelpfulBtn btn btn-outline-dark btn-sm"
                  disabled={false}
                  onClick={[Function]}
                  type="button"
                >
                  Yes (
                  5
                  )
                </button>
                <button
                  className="aReportBtn btn btn-outline-dark btn-sm"
                  disabled={false}
                  onClick={[Function]}
                  type="button"
                >
                  Report
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="moreAnswersRow row"
    >
      <button
        className="moreAnswers btn btn-outline-dark"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        See More Answers
      </button>
    </div>
    <div
      className="submitAnswerRow row"
    >
      <div
        className="submitAnswer"
      >
        <div
          className="submitAButton"
        >
          <button
            className="btn btn-outline-dark"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            Submit Answer
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
